showing alternative periods and edit of css when lessons are cancelled

ZuzkaL opened the pull request:
# Here are the results:
## Alternative periods for the current day plan
![image](https://github.com/zorzalo3/zorza/assets/26010258/907c8b6b-a5c4-4159-8a48-8c3a72923e4a)

## When there are cancelled lessons
![image](https://github.com/zorzalo3/zorza/assets/26010258/89119737-874b-479e-957b-e9f300533694)



# Git Issue: https://github.com/zorzalo3/zorza/issues/4

opti0 commented:
When a single lesson is canceled by the substitution module, no change should be made. Only a lesson cancellation added as a daily schedule for the entire school should add a strikethrough.

ZuzkaL commented:
Hi @opti0 , I am not sure what do you mean. Do you mean the case when there is only one lesson a day and it gets cancelled? What if there are more lessons and they all get cancelled? However, I would suggest creating another issue for that since it is a bit different from the original issue. Then I can look into it further.

opti0 commented on timetable/templates/substitutions.html:
Changes should only be visible after adding a day plan, not a substitution- cancellation, so this change is unnecessary.

ZuzkaL replied:
I was trying to implement, what you mentioned here: [https://github.com/zorzalo3/zorza/issues/4#issuecomment-1755620648](https://github.com/zorzalo3/zorza/issues/4#issuecomment-1755620648)

opti0 replied:
![obraz](https://github.com/zorzalo3/zorza/assets/105982904/b2d374a6-3c67-4813-ae83-ece6f92ed1ec)
when you add day plan -------- it displays information about the cancellation of all lessons. only then should this strikethrough appear. Adding a substitution - cancelation should not show this strikethrough.

ZuzkaL replied:
I have removed the strike through part

opti0 replied:
ok, now we wait for @Igifigi's review

Igifigi commented on timetable/utils.py:
```diff
@@ -44,11 +44,14 @@ def get_timetable_context(lessons):
     periods = [period.number for period in default_periods]
     period_strs = get_period_strings(default_periods)
 
+    alternative_periods = Period.objects.filter(schedule=2)
```
I don't think you should hardcode stuff like this. What if for example school will have 5 different schedules?

ZuzkaL replied:
Okay, I will take a look at it.

ZuzkaL replied:
@Igifigi I have fixed it so it works with multiple alternative schedules.

Igifigi commented on timetable/templates/substitutions.html:
```diff
@@ -28,7 +28,7 @@ <h2 class="text-center">{% trans "Substitutions" %}</h3>
 	<tr {% if group in groups or timetable_teacher != None and timetable_teacher == substitution.substitute %}class="relevant"{% endif %}>
 		<td>{{ substitution.lesson.period }}</td>
-		<td>{{ substitution.period_str }}</td>
+		<td {% if substitution.substitute == None %} style="text-decoration: line-through;"{% endif %}>{{ substitution.period_str }}</td>
```
The case for issue #4 is showing periods in the timetable part, not in the substitutions part below. Here you wrote that if one class has cancelled lesson, the period string in that row is crossed, which is nonesense, cause it's not the point of the issue

ZuzkaL replied:
I was trying to address https://github.com/zorzalo3/zorza/issues/4#issuecomment-1755620648 . However, I have already removed this change

Igifigi commented on timetable/templates/timetable_base.html:
```diff
@@ -43,9 +43,19 @@ <h3 class="timetable-caption">{% block caption %}{% endblock %}</h3>
-					{% autoescape off %}
-					{{ row.0|hyphen_break }}
-					{% endautoescape %}
+					<div style="color: #888888 ;">
+						<!--  Alternative date -->
+						{% if row.2 %} 
+							{% autoescape off %}
+							{{ row.2|hyphen_break }}
+							{% endautoescape %}
+						{% endif %}
+					</div>
+					<div>
+						{% autoescape off %}
+						{{ row.0|hyphen_break }}
+						{% endautoescape %}
+					</div>
```
This would work well if you passed the right data retrieved from the db, but I don't think generating empty divs is a good idea. Try not generating them if not necessary.

ZuzkaL replied:
Okay, I will take a look at it.

ZuzkaL replied:
@Igifigi I have fixed it so it creates divs only if there are alternative schedules left

ZuzkaL commented:
I have added the possibility of multiple alternative schedules. See PR description for examples.
Also I have removed the alternation of css when lessons are cancelled per @Igifigi [comment](https://github.com/zorzalo3/zorza/pull/69#discussion_r1392288831)

ZuzkaL commented:
Hi @Igifigi , can you look at this changes?

Igifigi commented on timetable/utils.py:
```diff
@@ -51,8 +51,10 @@ def get_timetable_context(lessons):
         table[period] = (period_strs[period], OrderedDict())
-        for alt_period in alternative_periods.filter(number=period):
-            table[period] = table[period] + (alt_period,)
+        for dayPlan in DayPlan.objects.all():
+            if dayPlan.is_today and not dayPlan.schedule.is_default:
+                for alt_period in alternative_periods.filter(number=period):
+                    table[period] = table[period] + (alt_period,)
```
When you've got more than one non-default schedule, the system displays them all.
![image](https://github.com/zorzalo3/zorza/assets/44808028/6c7198ec-a9bc-4cee-8b3f-1e484c300b57)
![image](https://github.com/zorzalo3/zorza/assets/44808028/d4cbeb9b-1c27-417d-aa0a-019b9baae377)

Igifigi replied:
There is no point in displaying multiple (more than default + alternative assigned for today) periods in the timetable. More details you can find in the issue. https://github.com/zorzalo3/zorza/issues/4#issuecomment-1838217936

Igifigi commented on timetable/utils.py:
```diff
@@ -48,13 +48,19 @@ def get_timetable_context(lessons):
     #       period number as key?
+    alternative_periods = Period.objects.exclude(schedule=1) # Exclude default periods
+    alternative_day_plan = None
+    for day_plan in DayPlan.objects.all():
+        if day_plan.is_today:
+            alternative_day_plan = day_plan
+            break
```
In most cases, if you have to use `break` in your code, something is poorly designed. Why are you iterating over all the DayPlans if you can just ask the database for a DayPlan for today in try-except block?

ZuzkaL replied:
I have simplified the query on DayPlan table as well, thanks for the suggestion!

Igifigi commented on timetable/utils.py:
```diff
@@ -48,13 +48,19 @@ def get_timetable_context(lessons):
-                for alt_period in alternative_periods.filter(number=period):
-                    table[period] = table[period] + (alt_period,)
+        if alternative_day_plan is not None:
+            alt_period = alternative_periods.filter(number=period, schedule=alternative_day_plan.schedule)[0]
```
Note that the tuple `(number, schedule)` is unique_together [here](https://github.com/zorzalo3/zorza/blob/master/timetable/models.py#L103), so using `filter` method is irrlevant. Additionally, you are wrongly assuming that for every period in default schedule there is equivalent period in alterantive schedule. For example during shortened lessons, the 9th and 10th lessons are cancelled (they are not included in the shortened schedule). You can cross out the periods that are cancelled (when `alternative_periods.get(number=period, schedule=alternative_day_plan.schedule)` throws `DoesNotExist` error). This is why I got IndexError - you tried to access the first element of empty list (`alternative_periods.filter(number=10, schedule=44)` is empty queryset).
![image](https://github.com/zorzalo3/zorza/assets/44808028/3446cb24-510d-4898-8554-11b748476bd7)
**Please test your changes before comitting them.**

Igifigi commented on timetable/utils.py:
```diff
@@ -41,26 +41,26 @@ def get_timetable_context(lessons):
-        if day_plan.is_today:
-            alternative_day_plan = day_plan
-            break
+    try:
+        alternative_day_plan = DayPlan.objects.get(date=date.today())
+        alternative_periods = Period.objects.filter(schedule=alternative_day_plan.schedule)
+    except ObjectDoesNotExist:
+        alternative_day_plan = None
```
Here you are catching `ObjectDoesNotExist`. In our case it should work _**perfectly fine**_, but for the future I strongly recommend using specified `Model.DoesNotExist` exception, because you have greater control over which model throws an exception. You can confidently use `ObjectDoesNotExist` exception when you're 100% sure which query will cause the exception. If somehow `Period.objects.filter()` method throws an exception (which I don't beileve it will, but still), your `alternative_day_plan` will be wrongly set to `None`. Additionaly, by using specified model exception, the code will be more readable and easy to understand for other devs.

ZuzkaL replied:
Thanks, I agree that this is clearer and easier to understand this way. I have changed it to catch `DayPlan.DoesNotExist` exception and also to catch `Period.DoesNotExist` exception bellow.

Igifigi commented on timetable/utils.py:
```diff
@@ -41,26 +41,26 @@ def get_timetable_context(lessons):
+        alternative_periods = Period.objects.filter(schedule=alternative_day_plan.schedule)
+    except ObjectDoesNotExist:
+        alternative_day_plan = None
 
```
Before geting the alternative periods, you should check if the alternative schedule is not the default one.
![image](https://github.com/zorzalo3/zorza/assets/44808028/f6d532ef-0fe3-4f66-adc3-7f12902fcfef)
![image](https://github.com/zorzalo3/zorza/assets/44808028/8aff8c82-7639-4527-8903-51c606c71104)

Igifigi commented on timetable/utils.py:
```diff
@@ -48,8 +48,11 @@ def get_timetable_context(lessons):
+        if alternative_day_plan.schedule.is_default:
+            alternative_day_plan = None
+        else:
+            alternative_periods = Period.objects.filter(schedule=alternative_day_plan.schedule)
```
What if `alternative_day_plan` says that lessons are cancelled? Note that `DayPlan.schedule` can be `None` [here](https://github.com/zorzalo3/zorza/blob/master/timetable/models.py#L70). Then you get `AttributeError`, because `NoneType` object has no attribute `is_default`. You should check whether `alternative_day_plan.schedule` is not `None`.
**And again: please test your changes before commiting!**

![image](https://github.com/zorzalo3/zorza/assets/44808028/dbc49539-2137-4a2e-aa46-6422d67763a6)

ZuzkaL replied:
Thanks @Igifigi, I did not realize that the option "---------" is a valid option. Fixed 
![image](https://github.com/zorzalo3/zorza/assets/26010258/fadb413c-ecf7-4679-ba7f-15c14a717744)

Igifigi reviewed (approved):
Used `get_object_or_none` method to make code more readable. Added comments.
